content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-reconf.md

```diff
@@ -0,0 +1,274 @@
+component configuration and avoid introducing typos and YAML indentation mistakes.
+{{< /warning >}}
+
+### Applying `ClusterConfiguration` changes
+
+#### Updating the `ClusterConfiguration`
+
+During cluster creation and upgrade, kubeadm writes its
+[`ClusterConfiguration`](https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/)
+in a ConfigMap in the `kube-system` namespace called `kubeadm-config`.
+
+To change a particular option in the `ClusterConfiguration` you can edit the ConfigMap with this command:
+
+```shell
+kubectl edit cm -n kube-system kubeadm-config
+```
+
+The configuration will be located under the `data.ClusterConfiguration` key.
+
+{{< note >}}
+The `ClusterConfiguration` includes a variety of options that affect the configuration of individual
+components such as kube-apiserver, kube-scheduler, kube-controller-manager, CoreDNS, etcd and kube-proxy.
+Changes to the configuration must be reflected on node components manually.
+{{< /note >}}
+
+#### Reflecting `ClusterConfiguration` changes on control plane nodes
+
+kubeadm manages the control plane components as StaticPod manifests located in
+the directory `/etc/kubernetes/manifests`.
+Any changes to the `ClusterConfiguration` under the `apiServer`, `controllerManager`, `scheduler` or `etcd`
+keys must be reflected in the associated files in the manifests directory on a control plane node.
+
+Such changes may include:
+- `extraArgs` - requires updating the list of flags passed to a component container
+- `extraMounts` - requires updated the volume mounts for a component container
+- `*SANs` - requires writing new certificates with updated Subject Alternative Names.
+
+Before proceeding with these changes make sure you backup the directory `/etc/kubernetes/`.
+
+To write new certificates you can use:
+```shell
+kubeadm init phase certs <component-name> --config <config-file>
+```
+
+To write new manifest files in `/etc/kubernetes/manifests` you can use:
+
+```shell
+kubeadm init phase control-plane <component-name> --config <config-file>
+```
+
+The `<config-file>` contents must match the updated `ClusterConfiguration`.
+The `<component-name>` value must be the name of the component.
+
+{{< note >}}
+Updating a file in `/etc/kubernetes/manifests` will tell the kubelet to restart the component StaticPod.
+Try doing these changes one node at a time to leave the cluster without downtime.
+{{< /note >}}
+
+### Applying kubelet configuration changes
+
+#### Updating the `KubeletConfiguration`
+
+During cluster creation and upgrade, kubeadm writes its
+[`KubeletConfiguration`](/docs/reference/config-api/kubelet-config.v1beta1/)
+in a ConfigMap in the `kube-system` namespace called `kubelet-config`.
+
+You can edit the ConfigMap with this command:
+
+```shell
+kubectl edit cm -n kube-system kubelet-config
+```
+
+The configuration will be located under the `data.kubelet` key.
+
+#### Reflecting the kubelet changes
```

Comment:
Can we just run `kubeadm upgrade node` after we edit the configmap? Is there any side-effect?

Reply:
i need to check.

Reply:
ok, so we can replace:
> - Edit the contents of the file `/var/lib/kubelet/config.conf` file to
match the contents of the `kubelet-config` ConfigMap

with:
> - Run `kubeadm upgrade node phase kubelet-config` to download the latest `kubelet-config` ConfigMap
> contents into the local file `/var/lib/kubelet/config.conf`

but we can keep the extra step for `/var/lib/kubelet/kubeadm-flags.env`, since it's not managed by `kubeadm upgrade` (unless a particular release changes something there - e.g. flag removal).